### DISCUSSION TITLE
Azure Open Telemetry - ActivityListener is null after first time

### DISCUSSION THREAD

**Discussion by tinuverma** [dotnet/runtime#83904] [Q&A] [answered]
I am trying Azure open telemetry and notice a weird behavior. First time all works well and data gets sent to telemetry. But any subsequent calls the ActivityListener is null and hence no data is sent. Can you please tell me what am I missing?

I have put all details in stackoverflow question here. Ignore title since I was still learning when I first put it there.
https://stackoverflow.com/questions/75817875/azure-open-telemetry-activity-is-null-few-min-after-app-start

Thanks
D

**Comment by tinuverma** [comment]
@tarekgh  any suggestions? You helped ma-gu with this question https://github.com/dotnet/runtime/issues/45070 ...so found your reference from there. Not sure how to tag ma-gu

**Comment by tarekgh** [comment, marked as answer]
@tinuverma 

You mentioned the following

```
I can confirm that the handleMessageActivity object is null and the activitySource still looks to be a valid object. Same problem happens on both activitySource?.CreateActivity and activitySource?.StartActivity I cannot understand why activitySource is valid object when the object is first deployed and then becomes null.
```

I am confused as you are saying `the activitySource still looks to be a valid object` and at the end of text you are saying `I cannot understand why activitySource is valid object when the object is first deployed and then becomes null`. 
Does the `activitySource` is valid and not null when the problem occurs or not? 

If it is valid, can you try to call `activitySource.HasListeners()` and find out what it will return. My suspicion here is the `activitySource` is disposed. 

C @noahfalk

**Reply by tinuverma** [reply]
@tarekgh  @noahfalk 
this was work in progress so some of the information may  be confusing. Here is the time line
I start the app.
Publish test message 1
I see data in telemetry - clearly all objects are setup the way they should be
I wait 2-5 min
Publish test message 1 again
No data is published to telemetry
When I watch the objects in handler (client) class, here is what I notice 
"activitySource " is found by the "GetActivitySource" method
in activitySource there is a property HasListeners - that is now null.
So what I am trying to figure out is why is it null.

If activitySource is getting disposed, then on second call it should have been null - but it's not.
Does that help? 
Thank

**Reply by tinuverma** [reply]
I was unable to upload screen shots here ..so upload the screen shots to stackoverflow question.

**Reply by tarekgh** [reply]
> in activitySource there is a property HasListeners - that is now null.

`HasListeners` is a method and not a property on the `ActivitySource`. This method returns a Boolean value `true` or `false`.  Could you try to call `activitySource.HasListeners()` before and after the problem happens and tell if this return true or false?

**Reply by tinuverma** [reply]
when the problem occurs,  activitySource.HasListeners() returns false.
In the screen shot I have shown the _listeners is null.

**Reply by tarekgh** [reply]
> when the problem occurs, activitySource.HasListeners() returns false.

This means the ActivitySource object somehow is disposed, or the listeners used to attach to it is disposed. it may be helpful if you can set a breaking point on the `ActivitySource.Dispose` and run your scenario and find out if the debugger will stop there. This can give cluse of the stack which disposed the object.

**Reply by tinuverma** [reply]
Both ActivitySource and listener classes are sealed. How would I hook into their dispose method?

**Reply by tinuverma** [reply]
and dispose method in my code did not get called

**Reply by tarekgh** [reply]
Is it possible you can create a small repro and share it? I may try to help debug it.

**Reply by tinuverma** [reply]
I was able to figure out one thing...since my Handle code is sending telemetry data with a "using block", the said activity source object most likely is getting disposed after the first message is handled. 
I changed my service to return a new ActivitySource object everytime and now "HasListeners" returns true.
So it looks like the solution is to not create the AcitivitySource initially but on demand.
How bad is the performance hit with this approach? Or am I missing something here?
Updated this on stackoverflow as well.

**Reply by tarekgh** [reply]
Are you creating a new ActivitySource every time or are you saying you are creating it lazily? if the former, then that will not be good for the perf especially if the system has some enabled listeners. try to get the ActivitySource out of the using block and cache it statically.

**Reply by tinuverma** [reply]
Once I did that then I see logs being written but those logs dont make much sense since all. Here is the end result I see right now:
if 5 messages are handled, I see 5 "ActivitySources" which is a good thing. But under each I see the same set of "Activities". So if first message should put 3 activities, then I expect to see three activities under ActivitySource one handler but I see 15 since they are all getting clubbed together.
I looked for some way to stop, not dispose, activitySource after a message is finished being handled, but I dont see any.
Do you have any suggestions on how they should be linked so the activity does not spill over?
Thanks
D

**Reply by tinuverma** [reply]
it could also be because of link at wrong place. looking now
wait..Ignore since it's inside using and I think is not a problem...must be something else.

**Reply by tarekgh** [reply]
> I see 5 "ActivitySources" which is a good thing.

Creating `ActivitySource` per request is expensive. Can't you create a global static ActivitySource to handle that? 

> But under each I see the same set of "Activities"

This cannot be. You can check `Activity.Source` to know which ActivitySource created such Activity.  Try to set a break point where you create Activities to know what is going on.

**Reply by tinuverma** [reply]
Creating ActivitySource per request is expensive. Can't you create a global static ActivitySource to handle that?

Look at the code in stackoverflow - it is created once only. Since I do not dispose it, the same object should get reused.

In my foreach loop, I am starting an activity. that's what I mean. the parent object has a collection of child objects. I am creating a child activity for each item in the list

`foreach (var personId in message.SomeList.Select(a => a.personId)
                .Distinct())
            {
                using (var processPartyActivity = activitySource?
                    .StartActivity(kind: ActivityKind.Client, 
                    links: links, name: "process party"))
                {....`

**Reply by tarekgh** [reply]
Can you summarize what is the current issue you are trying to solve after your discoveries?

**Reply by tinuverma** [reply]
I rearranged the code and now it looks so much cleaner - the logs I mean. Here is the summary of changes I made. I will put the code in stackoverflow link for others also.

- created ActivitySource once on startup for each "source" that I want to capture telemetry on
- in my handler, I get the ActivitySource I want to work with ...no using statement at this level to avoid dispose getting called on it
- in a using statement, started activity for message I receive using activitySource?.StartActivity($"some constant name", ActivityKind.Internal)
- in my message there is a list of objects ...so I created a second activity and linked it to the above one.

This fixed the issues I was having. 
I used "StartActivity" instead of create activity so I dont have to manually start and stop it (one of my bugs)
when I created second activity for the list of objects i linked it to parent activity using the method processPartyActivity.SetParentId(handleMessageActivity?.Id);
This fixed hierarchy issue.

Thanks much for your help.

**Reply by tinuverma** [reply]
@tarekgh  in the application map, I noticed it's not drawing the dependency correctly. Keeping in mind the code I shared in stackoverflow link, I expected from the base component a line to handleMessageActivity and then a second line to processPartyActivity  ...instead I see line from base to handleMessageActivity and from base again to ProcessPartyActivity . Looks like I am missing something in the relationship. Can you please see if you find the mistake?

**Reply by tarekgh** [reply]
Looks to me you are creating the activity `handleMessageActivity` and not starting it.

**Reply by tinuverma** [reply]
I did notice that and in the comment I marked as answered, I did mention to change handleMessageActivity  to StartActivity instead of CreateActivity. From what I read in docs, StartActivity would automatically start it and since it's inside using statement automatically stop and dispose it. I do see the output from handleMessageActivity   in telemetry which confirms the code is working. Just the workflow is not right.

**Reply by tarekgh** [reply]
Can you paste the latest code here and tell what you are getting and what you expect to get? This may make it easier to follow up with the discussion.

**Reply by tinuverma** [reply]
```C#
var activitySource = _azureOpenTelemetryLoggerService.GetActivitySource(AzureOpenTelemetryService.ActivitySourceEnum.ProductAgreementRecordChangeHandler);

using (var handleMessageActivity = activitySource?.StartActivity($"Handle ProductAgreementRecordChange message", ActivityKind.Internal))
{
	handleMessageActivity?.SetTag("Prop 1", message.Prop1);
	handleMessageActivity?.SetTag("Prop 2", message.Prop2);
	ActivityContext? linkedContext = handleMessageActivity?.Context;
	var links = new List<ActivityLink>();
	if (linkedContext.HasValue)
	{
		links.Add(new ActivityLink(linkedContext.Value));
	}
	foreach (var p_id in message.PersonsList.Select(a => a.PersonId).Distinct())
	{
		using (var processPartyActivity = activitySource?.StartActivity(kind: ActivityKind.Client, links: links, name: "process person"))
		{
			if (handleMessageActivity != null)
			{
				processPartyActivity.SetParentId(handleMessageActivity?.Id);
			}
			processPartyActivity?.AddTag("personId", personId);
			processPartyActivity?.AddEvent(new ActivityEvent($"Call person service for  {p_id}"));
			// do stuff
		}
	}
}
```

**Reply by tinuverma** [reply]
in workflow I expected to see **Message enters** --> **handleMessageActivity**  --> **processPartyActivity**  ...but I see **Message enters** --> **handleMessageActivity**   and **Message enters** --> **processPartyActivity**

**Reply by tarekgh** [reply]
try to change 

```diff
-    using (var processPartyActivity = activitySource?.StartActivity(kind: ActivityKind.Client, links: links, name: "process person"))
+    using (var processPartyActivity = activitySource?.StartActivity("process person", ActivityKind.Client, linkedContext.Value, tags: null, links))
```

and see if this will help? 

Also, doing 

```C#
if (handleMessageActivity != null)
			{
				processPartyActivity.SetParentId(handleMessageActivity?.Id);
			}
```

is wrong, you cannot set the parent after the activity starts. Delete these lines.

**Reply by tinuverma** [reply]
I will try the change to how processPartyActivity  is initialized and try.
How would I set processPartyActivity  as child of handleMessageActivity then? Or does it automatically happen because of the nesting?

**Reply by tarekgh** [reply]
> How would I set processPartyActivity as child of handleMessageActivity then?

In `activitySource?.StartActivity` I am passing the `linkedContext` as the parent context.

**Reply by tinuverma** [reply]
while what you said made sense, the workflow did not change. it's still showing it incorrect.

**Reply by tarekgh** [reply]
Why do you pass the `links` too? if we are setting the parent context correctly, you don't need to pass the link, right? can you try to remove it and look if this make any difference?

**Reply by tinuverma** [reply]
didnt make any difference

**Reply by tarekgh** [reply]
> I see Message enters --> handleMessageActivity and Message enters --> processPartyActivity

Where do you see these messages? is it through OpenTelemetry or something else?

**Reply by tinuverma** [reply]
sent screen shot in email to you. if not inbox, please look at spam folder.

**Reply by tarekgh** [reply]
Yes, I got the email. 

Can you add console exporter and send its output too? you may look at https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/getting-started-console/README.md for how to do it.

**Comment by tarekgh** [comment]
@tinuverma please log issue in https://github.com/Azure/azure-sdk-for-net/issues and they should be able to help you with the Azure monitor. From your code now, the activity relationship is set correctly.
